### ISSUE TITLE
SSH invalid argument error with TeamCity since 3.x

### ISSUE THREAD

**Issue by chrisdjali-wrld3d** [git-lfs/git-lfs#4733]
**Describe the issue**
I imagine you want to class this as TeamCity's fault instead of a regression in Git LFS, but it's more discoverable here, and possibly you want to restore compatibility, so I'm posting this here anyway.

When using TeamCity for CI, if it's configured to do agent-side checkout (a requirement for Git LFS support) and a key's been set in the VCS root settings, at least in the 2017 version (I don't have access to anything up-to-date, so I can't verify whether anything's been changed), it applies the key by directing Git to use a temporary batch file as its SSH command, which presumably adds the necessary arguments and forwards the command to the real SSH.

Git LFS 2.x would call SHH with commands of the form `ssh arg arg arg`, whereas 3.x calls it as `ssh -- arg arg arg` (i.e. with `--`). It appears TeamCity's shim doesn't support this, as the following error is emitted:

```
Error downloading object: path (short hash): Smudge error: Error downloading path (long hash): batch request: Invalid arguments [--, git@github.com, git-lfs-authenticate org/repo.git download]: exit status 1
```

Or, in the `GIT_TRACE` file:
```
12:08:50.384288 trace git-lfs: run_command: C:\buildAgentData\temp\git-ssh-3502009113789195714.bat -- git@github.com git-lfs-authenticate org/repo.git download
12:08:50.384860 trace git-lfs: exec: C:\buildAgentData\temp\git-ssh-3502009113789195714.bat '--' 'git@github.com' 'git-lfs-authenticate org/repo.git download'
12:08:50.515545 trace git-lfs: ssh: git@github.com failed, error: exit status 1, message: Invalid arguments [--, git@github.com, git-lfs-authenticate org/repo.git download] (try: 0/6)
```

Comparing `GIT_TRACE` output between this and the same command from the last working version shows the command given to the SSH shim batch file is identical except for the absence of the `--`.

This isn't another case of #4672 as there's no leading slash in the org name.

It's definitely the shim batch file that's triggering the problem as running the same commands as TeamCity does, but manually, allows the checkout to work and the normal SSH command is used instead of the batch file

**System environment**
Windows Server 2019 Datacenter version 1809 build 17763.2237
Git called via TeamCity Enterprise 2017.2.2 (build 50909) with the agent process running as a service.
Git for Windows 2.33.1.windows.1
Git LFS 3.0.1

**Output of `git lfs env`**
This is likely inaccurate as I can't run this in the same context as Git runs when things break and I expect that some environment variables are different.
```
git-lfs/3.0.1 (GitHub; windows amd64; go 1.17.1)
git version 2.33.1.windows.1

Endpoint=https://github.com/org/repo.git/info/lfs (auth=none)
  SSH=git@github.com:org/repo.git
LocalWorkingDir=C:\buildAgentData\work\c43e8748fe05225d\repo
LocalGitDir=C:\buildAgentData\work\c43e8748fe05225d\repo\.git
LocalGitStorageDir=C:\buildAgentData\work\c43e8748fe05225d\repo\.git
LocalMediaDir=C:\buildAgentData\work\c43e8748fe05225d\repo\.git\lfs\objects
LocalReferenceDirs=
TempDir=C:\buildAgentData\work\c43e8748fe05225d\repo\.git\lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=C:\buildAgentData\work\c43e8748fe05225d\repo\.git\lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
GIT_TRACE=C:\git-trace.log
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**
As mentioned right at the top, this is really TeamCity's fault for making an SSH shim that can't handle all the arguments that might be passed to SSH, not Git LFS' fault for passing SSH arguments SSH is compatible with.

**Comment by KalleOlaviNiemitalo** [comment]
Perhaps setting GIT_SSH_VARIANT=simple in the environment or the equivalent `ssh.variant` in Git configuration would make it work.

**Comment by chrisdjali-wrld3d** [comment]
Apparently not - it still passes the `--`.

**Comment by KalleOlaviNiemitalo** [comment]
I tried a few variations of `GIT_TRACE=1 GIT_SSH_VARIANT=ssh git ls-remote origin` and didn't see Git 2.33.0.windows.2 itself passing `--` to ssh. I don't see anything related in <https://github.com/git/git/blob/v2.33.1/Documentation/RelNotes/2.33.1.txt>, either. In my opinion, it's a bug in Git LFS if GIT_SSH_VARIANT=simple doesn't make it use the same options as Git would.

(However, fixing this might not solve the problem for TeamCity setups because "ssh variant 'simple' does not support setting port".)

**Comment by KalleOlaviNiemitalo** [comment]
Git LFS does check the SSH variant, though:

<https://github.com/git-lfs/git-lfs/blob/4f926516b20df86928c894a52fe5e39fe9c58bcf/ssh/ssh.go#L176-L188>

I wonder why that didn't work in your test. Might TeamCity itself override the GIT_SSH_VARIANT setting?

**Comment by chrisdjali-wrld3d** [comment]
> I tried a few variations of `GIT_TRACE=1 GIT_SSH_VARIANT=ssh git ls-remote origin` and didn't see Git 2.33.0.windows.2 itself passing `--` to ssh. I don't see anything related in https://github.com/git/git/blob/v2.33.1/Documentation/RelNotes/2.33.1.txt, either. In my opinion, it's a bug in Git LFS if GIT_SSH_VARIANT=simple doesn't make it use the same options as Git would.
> 
> (However, fixing this might not solve the problem for TeamCity setups because "ssh variant 'simple' does not support setting port".)

2.33.0.windows.1 and 2.33.0.windows.2 use Git LFS 2.x - 2.33.1.windows.1 is the first that uses Git LFS 3.x. You need to look at the changelog for Git for Windows rather than upstream Git: https://github.com/git-for-windows/git/releases/tag/v2.33.1.windows.1.

> Might TeamCity itself override the GIT_SSH_VARIANT setting?

This is possible - I can't make it run arbitrary commands during its checkout process, only during the build itself, and it doesn't meddle with the environment variables in the same way for the two steps. I don't see anything in my `GIT_TRACE` log indicating this one way or the other (which makes sense as there're no relevant `tracerx` calls in the file you linked).

**Comment by KalleOlaviNiemitalo** [comment]
> 2.33.0.windows.1 and 2.33.0.windows.2 use Git LFS 2.x - 2.33.1.windows.1 is the first that uses Git LFS 3.x. 

Irrelevant because I was testing whether Git itself adds `--`.

**Comment by KalleOlaviNiemitalo** [comment]
On Windows, you could use [Process Monitor](https://docs.microsoft.com/sysinternals/downloads/procmon) to check the environment variables received by Git and by the SSH wrapper script.

**Comment by chrisdjali-wrld3d** [comment]
Of course! I even did that earlier in this investigation to verify I was setting `GIT_TRACE` (although actually I used Process Explorer instead as accidentally leaving procmon running in the background and then having my machine lock up due to it overflowing memory and using the page file has bitten me enough times that I use it sparingly).

The Git-related environment variables during the clone are:
```
GIT_CONFIG_PARAMETERS	'credential.helper'=''
GIT_EXEC_PATH	C:/Program Files/Git/mingw64/libexec/git-core
GIT_PREFIX	
GIT_SSH	C:\buildAgentData\temp\git-ssh-3373411814272472892.bat
GIT_TRACE	C:\git-trace.log
GIT4IDEA_SSH_HANDLER	372865295
GIT4IDEA_SSH_IGNORE_KNOWN_HOSTS	true
GIT4IDEA_SSH_PORT	9090
```

It appears that the method I used to set `GIT_SSH_VARIANT` only gets applied during the build step, not the checkout step, so didn't actually do anything.

When `GIT_SSH_VARIANT` is actually passed to the process correctly, it does fix the problem. However, I'm not sure that's a viable workaround in general - some of our CI jobs call Git directly, in which case TeamCity isn't intercepting SSH calls, so we might want the variant autodetection working.

**Comment by KalleOlaviNiemitalo** [comment]
A hacky idea: replace `git` with a wrapper that checks whether `GIT_SSH` is a TeamCity wrapper and adds `GIT_SSH_VARIANT` at that time.

Has this incompatibility been reported to JetBrains?

**Comment by chrisdjali-wrld3d** [comment]
It hasn't yet as we're stuck on an older version that's definitely out of support. As I mentioned in my original post, it's possible they've already fixed it in a version we don't have access to.

**Comment by bk2204** [comment]
Hey,

We've seen another report of this in #4705.  As I mentioned there, this is required for security.  However, since it seems to break things in a lot of cases, I'll try to get a fix in for 3.1.0 which uses `--` less often.  It may still come up in some cases, though, so TeamCity should take care to implement this properly.

**Comment by chrisdjali-wrld3d** [comment]
I'm surprised I missed that. I had a pretty thorough look for an existing ticket before opening this one.
